google/cloud/bigquery/table.py

```diff
@@ -1011,6 +1011,13 @@ def _build_resource(self, filter_fields):
         """Generate a resource for ``update``."""
         return _helpers._build_resource_from_properties(self, filter_fields)
 
+    def __eq__(self, other):
+        if not isinstance(other, Table):
```

Comment:
Should we allow TableReference and TableListItem as well?

Reply:
Hm... since we only care about table reference and ignore other metadata properties, and because `Table` and `TableReference` can already be used interchangeably as an argument to many methods, this could actually make sense.

Reply:
Done.

However, I now see quite a lot of repetition - all these `Table*` classes have the same `__hash__()` definition, the equality comparison is very similar, and they all have the same core properties - `project`, `dataset_id`, `table_id`.

This seems like a good refactoring opportunity - I would extract this into a common base class. What do you think?

Reply:
Yes, a common base class would make sense. I think there are a few other methods / properties that are duplicated besides `__hash__`. Let's do that in another PR just for the refactoring.